### PR TITLE
Multiple targets

### PR DESCRIPTION
**Pull request by irina060981**
For the issue #14 , https://github.com/alpheios-project/alignment-editor-new/issues/37, https://github.com/alpheios-project/alignment-editor-new/issues/38

1. Added feature for having multiple targets - add target, delete target, add targetId condition for various methods
2. Added limits for starting align workflow:
    - the amount of segments should be equal
    - all texts should be fully defined
    - croos-segments groups are not allowed
3. Added Vuex for handling observable changes
4. Removed an ability to change source after align workflow started
5. Changed html layout for showing aligning segments as a table-view
6. Added tabs for multiple target texts = edit, view mode
7. Added descriptions and tests

Updated working link:

https://alpheios-misc-dev.s3.us-east-2.amazonaws.com/alignment-editor/index.html

Some screenshots:

image

image

image

image

image

### CODE REVIEW

**Review by kirlat — CHANGES_REQUESTED** [review]
It looks good, I have just a few minor suggestions. I might miss some finer business logic details because of a sheer volume, but I've tried my best 🙂. I like the clean UI of the editor and the way code is organized.

**Review by kirlat — CHANGES_REQUESTED** [review]
It looks good, I have just a few minor suggestions. I might miss some finer business logic details because of a sheer volume, but I've tried my best 🙂. I like the clean UI of the editor and the way code is organized.

File: src/lib/data/alignment.js [hunk 1]
```diff
@@ -374,13 +425,11 @@ export default class Alignment {
    * @param {Token} token
    * @returns {Boolean} true - groups were merged, false - was not
    */
-  mergeActiveGroupWithAnotherByToken (token) {
-    if (this.hasActiveAlignment && this.tokenIsGrouped(token)) {
-      const tokensGroup = this.findAlignmentGroup(token)
+  mergeActiveGroupWithAnotherByToken (token, limitByTargetId) {
+    if (this.hasActiveAlignmentGroup && !this.tokenInActiveGroup(token, limitByTargetId) && this.tokenIsGrouped(token, limitByTargetId)) {
```
Comment (balmas) [inline src/lib/data/alignment.js:429]: There is a bug here (i think). The segment is not checked when merging two groups. To reproduce the bug: 

click origin text token A Segment 1
click target text token B Segment 1
click origin text token A Segment 1(alignment group is saved)
click origin text token C Segment 2
click origin text token A Segment 1
click origin text token C Segment 2
The alignment group now contains Origin text Token A from Segment 1 Token C from Segment 2 and Target Text Token B from Segment 1

File: src/vue/align-editor/align-editor.vue [hunk 2]
```diff
@@ -1,51 +1,35 @@
 <template>
   <div class="alpheios-alignment-editor-container" v-show="showAlignEditor">
+    <div class="alpheios-alignment-editor-header">
```
Comment (balmas) [inline src/vue/align-editor/align-editor.vue:3]: the ALIGN_EDITOR_HEADING message is wrong. It says "Define Origin and Target texts" (i.e. same as the text editor heading

File: src/vue/main-menu.vue [hunk 3]
```diff
@@ -1,8 +1,21 @@
 <template>
   <div class="alpheios-alignment-app-menu" id="alpheios-main-menu">
       <div class="alpheios-alignment-app-menu__buttons">
-        <button class="alpheios-button-tertiary" id ="alpheios-main-menu-download" @click="$emit('download-data')" >{{ l10n.getMsgS('MAIN_MENU_DOWNLOAD_TITLE') }}</button>
-        <button class="alpheios-button-tertiary" id ="alpheios-main-menu-upload" @click="uploadTexts" >{{ l10n.getMsgS('MAIN_MENU_UPLOAD_TITLE') }}</button>
+        <button class="alpheios-button-tertiary" id ="alpheios-main-menu-add-target" 
+                @click="$emit('add-target')" :disabled="!docSourceEditAvailable" >
```
Comment (balmas) [inline src/vue/main-menu.vue:5]: eventually we want to be able to add target texts once the alignment has already started. Ok for this to be in a separate PR but can you make an issue so we don't forget?

File: src/vue/main-menu.vue [hunk 4]
```diff
@@ -1,8 +1,21 @@
...
+        <button class="alpheios-button-tertiary" id ="alpheios-main-menu-download" 
+                @click="$emit('download-data')"  :disabled="!docSourceEditAvailable" >
+                {{ l10n.getMsgS('MAIN_MENU_DOWNLOAD_TITLE') }}
+        </button>
+
+        <button class="alpheios-button-tertiary" id ="alpheios-main-menu-upload" 
+                @click="uploadTexts"  :disabled="!docSourceEditAvailable" >
+                {{ l10n.getMsgS('MAIN_MENU_UPLOAD_TITLE') }}
+        </button>
+
```
Comment (balmas) [inline src/vue/main-menu.vue:18]: I think we need to add a handler to catch the selectTab event sent from the align-editor-tabs component so that Undo and Redo can also be disabled when in View Only mode. It doesn't make sense for them to be  active while click is disabled does it?

**Review by balmas — CHANGES_REQUESTED** [review]
nice work. A few changes needed but on the whole I think the approach is working.

File: src/lib/controllers/aligned-controller.js [hunk 5]
```diff
@@ -11,23 +15,54 @@ export default class AlignedController {
       console.error(L10nSingleton.getMsgS('ALIGNED_CONTROLLER_NOT_READY_FOR_TOKENIZATION'))
       return false
     }
+
     this.alignment = alignment
     const tokenizer = 'simpleWordTokenization'
-    return this.alignment.createAlignedTexts(tokenizer)
+    const result = this.alignment.createAlignedTexts(tokenizer)
+
+    const res2 = this.alignment.equalSegmentsAmount
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/controllers/aligned-controller.js:23]: It is not possible to do this check first, the workflow here is the following:
1. We have some source texts (one origin -> multiple targets) similiar to (origin-> translations to different languages).
2. Tokenization service will define segments for all texts on createAlignedTexts steps:
    - we would send each source text to the tokenization service
    - we would receive back aligned texts that consist of segments
   The tokenization service doesn't return only segments amount.
3. On this step we could do the check of segments amount and reject tokenization results.

We discussed this steps order on the previous meeting and decided to go this way.

File: src/lib/controllers/aligned-controller.js [hunk 6]
```diff
@@ -11,23 +15,54 @@ export default class AlignedController {
...
     this.alignment = alignment
     const tokenizer = 'simpleWordTokenization'
-    return this.alignment.createAlignedTexts(tokenizer)
+    const result = this.alignment.createAlignedTexts(tokenizer)
+
+    const res2 = this.alignment.equalSegmentsAmount
+
+    if (!res2) {
+      console.error(L10nSingleton.getMsgS('ALIGNED_CONTROLLER_NOT_EQUAL_SEGMENTS'))
+      this.alignment.clearAlignedTexts()
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/controllers/aligned-controller.js:27]: We could have here two workflows on segments equality fail:
  - clear everything
  - a user could change any source text he/she likes , we don't track changes
   - align again from scratch on Align button click

or 
   - clear only wrong targets texts (because we should choose some of them that we consider to be correct, by default it could be an origin)
   -  freeze correct origin and targets and allow a user to change only the text that differs by segments amount
   - align again (and may be return again to the start)
or
   - clear only wrong targets texts
   - don't freeze and track changes for all texts 
   - if any changes would be done - clear aligned texts
   - align again (and may be return again to the start)

From my point of view the first one is the most easier and consistent.
If we want to reduce tokenization re-done, than it would be better to add simple segments checks to the tokenization service.
@balmas , if you think that some of the worklows is better - please tell me and I will change it (may be it is better to create a specific issue for that and do it through a new PR)? what do you think?

File: src/lib/controllers/aligned-controller.js [hunk 7]
```diff
@@ -11,23 +15,54 @@ export default class AlignedController {
...
+      this.store.commit('incrementAlignmentUpdated')
+      return false
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/controllers/aligned-controller.js:29]: @kirlat , I didn't finally understand what do you mean here saying Js constructor?
This is a method:
```
  createAlignedTexts (alignment) {
    if (!alignment || !alignment.readyForTokenize) {
      console.error(L10nSingleton.getMsgS('ALIGNED_CONTROLLER_NOT_READY_FOR_TOKENIZATION'))
      return false
    }

    this.alignment = alignment
    const tokenizer = 'simpleWordTokenization'
    const result = this.alignment.createAlignedTexts(tokenizer)

    const res2 = this.alignment.equalSegmentsAmount

    if (!res2) {
      console.error(L10nSingleton.getMsgS('ALIGNED_CONTROLLER_NOT_EQUAL_SEGMENTS'))
      this.alignment.clearAlignedTexts()
      this.store.commit('incrementAlignmentUpdated')
      return false
    }
    this.store.commit('incrementAlignmentUpdated')
    return result
  }
```
And it will always return Boolean value, true - if texts were aligned successfully, false - aligned texts were not created

File: src/lib/controllers/aligned-controller.js [hunk 8]
```diff
@@ -40,42 +75,47 @@ export default class AlignedController {
    * If there is an active alignment and token is not grouped - addToAlignmentGroup
    * @param {Token} token
    */
-  clickToken (token) {
-    if (!this.hasActiveAlignment) {
-      if (this.tokenIsGrouped(token)) {
-        this.activateGroupByToken(token)
+  clickToken (token, limitByTargetId) {
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/controllers/aligned-controller.js:78]: oh, yes, I missed to update it. Will fix

File: src/lib/controllers/history-controller.js [hunk 9]
```diff
@@ -35,15 +39,16 @@ export default class HistoryController {
...
-      return this.alignment.activateGroupByGroupIndex(this.alignment.alignmentGroups.length - 1)
+    let result
+    if (this.alignment.hasActiveAlignmentGroup && this.alignment.activeAlignmentGroup.groupLen > 1) {
+      result = this.alignment.undoInActiveGroup()
+    } else if (this.alignment.hasActiveAlignmentGroup && this.alignment.activeAlignmentGroup.groupLen === 1) {
+      result = this.alignment.undoActiveGroup()
+    } else if (!this.alignment.hasActiveAlignmentGroup && this.alignment.alignmentGroups.length > 0) {
+      result = this.alignment.activateGroupByGroupIndex(this.alignment.alignmentGroups.length - 1)
     }
+    this.store.commit('incrementAlignmentUpdated')
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/controllers/history-controller.js:50]: I am not sure that there could be such cases, because if we could do nothing this action won't be executed.
But I think that it would be good to add a condition to limit store.commit execution.
Thank you for the note.

File: src/lib/controllers/history-controller.js [hunk 10]
```diff
@@ -53,14 +58,17 @@ export default class HistoryController {
...
-      return this.alignment.returnActiveGroupToList()
+    if (this.alignment.hasActiveAlignmentGroup && this.alignment.currentStepOnLastInActiveGroup && this.alignment.undoneGroups.length > 0) {
+      result = this.alignment.finishActiveAlignmentGroup()
     }
-    if (!this.alignment.hasActiveAlignment && this.alignment.undoneGroups.length > 0) {
-      return this.alignment.redoActiveGroup()
+    if (!this.alignment.hasActiveAlignmentGroup && this.alignment.undoneGroups.length > 0) {
+      result = this.alignment.redoActiveGroup()
     }
+    this.store.commit('incrementAlignmentUpdated')
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/controllers/history-controller.js:71]: Would fix

File: src/lib/data/alignment-group.js [hunk 11]
```diff
@@ -42,15 +45,49 @@ export default class AlignmentGroup {
...
    * Adds the token to group, adds to step and checks firstStep
    * @param {Token} token
    * @returns { Boolean } true - token was added, false - not
    */
   add (token) {
-    if (!token || !token.isAlignable) {
+    if (!token || !token.isAlignable || !this.couldBeIncluded(token)) {
       return false
     }
+    if (this.groupLen === 0) {
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/data/alignment-group.js:85]: I think you are right, it is not really a correct check.
I would change the condition to the following:
```
    if (!this.segmentIndex) {
      this.segmentIndex = token.segmentIndex
    }
```

and I would an oposite condition to remove, to clear segmentIndex if we remove the last token to remove method

File: src/lib/data/token.js [hunk 12]
```diff
@@ -1,14 +1,20 @@
 export default class Token {
-  constructor ({ textType, idWord, word, beforeWord, afterWord, hasLineBreak }) {
+  constructor ({ textType, idWord, word, beforeWord, afterWord, hasLineBreak } = {}, segmentIndex, docSourceId) {
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/data/token.js:2]: Because it is created from the tokenization service,
we receive token-like object from the service, and pass it as the first parameter, and add aditional parameters from the environment, that were not used for the tokenization - `docSourceId` and `segmentIndex`.
So that's why 
`{ textType, idWord, word, beforeWord, afterWord, hasLineBreak }` - is main token data
`segmentIndex, docSourceId` - is additional data for optimizing perfomance

File: src/vue/align-editor/align-editor-view-mode.vue [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+     * Changes active tabs by click
+     * @param {String}
+     */
+    selectTab (targetId) {
+      if ((this.shownTabs.length > 1) && this.shownTabs.includes(targetId)) {
+        this.shownTabs = this.shownTabs.filter(innerTargetId => innerTargetId !== targetId)
+      } else if (!this.shownTabs.includes(targetId)) {
+        this.shownTabs.push(targetId)
+      } else {
+        return
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/vue/align-editor/align-editor-view-mode.vue:126]: you are right

File: src/vue/align-editor/align-editor.vue (hunk 2 above)
Comment (irina060981) [reply src/vue/align-editor/align-editor.vue:3]: Changed to "Define Alignment Groups"

File: src/vue/main-menu.vue (hunk 3 above)
Comment (irina060981) [reply src/vue/main-menu.vue:5]: It is registered here https://github.com/alpheios-project/alignment-editor-new/issues/31

I postponed it to the step, when I would work on edit alignment mode

File: src/vue/main-menu.vue (hunk 4 above)
Comment (irina060981) [reply src/vue/main-menu.vue:18]: oh, it is an interesting case, will add it

File: src/lib/controllers/aligned-controller.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (kirlat) [reply src/lib/controllers/aligned-controller.js:29]: Oh, sorry, I was confused by the way the code changes were represented by the GitHub and I thought this code was part of a constructor. Now I realize that I was wrong. So please ignore my comment, there is absolutely no issue with the return value then.

File: src/lib/data/token.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (kirlat) [reply src/lib/data/token.js:2]: That makes sense. Thanks for explaining!

File: src/lib/controllers/aligned-controller.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (kirlat) [reply src/lib/controllers/aligned-controller.js:23]: Thanks for explaining! If it's the best we can do then I'm fine with it.

File: src/lib/controllers/texts-controller.js [hunk 14]
```diff
@@ -33,18 +38,34 @@ export default class TextsController {
     } else {
       this.alignment.updateOriginDocSource(originDocSource)
     }
+    this.store.commit('incrementAlignmentUpdated')
   }
 
   /**
    * Uploads target source document to the alignment object.
    * If an alignment is not created yet, it would be created.
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/controllers/texts-controller.js:46]: Done

File: src/lib/data/alignment.js [hunk 15]
```diff
@@ -65,21 +79,41 @@ export default class Alignment {
...
-   * @param {SourceText} docSource
+   * @param {SourceText | Object} docSource
+   * @returns {Boolean}
    */
-  updateTargetDocSource (docSource) {
+  updateTargetDocSource (docSource, targetId) {
     if (!this.origin.docSource) {
-      console.error(L10nSingleton.getMsgS('ALIGNMENT_ERROR_ADD_TO_ALIGNMENT'))
+      if (docSource) {
+        console.error(L10nSingleton.getMsgS('ALIGNMENT_ERROR_ADD_TO_ALIGNMENT'))
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/data/alignment.js:89]: At the start of the application we have an empty text blocks without docSource.
I've added this check to prevent from printing an error in this case (on application init)

File: src/lib/data/alignment-group.js [hunk 16]
```diff
@@ -369,105 +420,4 @@ export default class AlignmentGroup {
 
     return finalResult
   }
-
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/data/alignment-group.js:372]: yes, I optimised code here, now it is more generic

File: src/lib/data/alignment.js (hunk 1 above)
Comment (irina060981) [reply src/lib/data/alignment.js:443]: Great, thank you for the catch!

Comment (irina060981) [reply src/lib/data/alignment.js]: Fixed

**Comment by irina060981** [comment]
@balmas , @kirlat  - I have worked through all notes, thank you for the attentive look to my code :)
Do you think I could merge?

**Comment by kirlat** [comment]
> @balmas , @kirlat - I have worked through all notes, thank you for the attentive look to my code :)
> Do you think I could merge?

I have no objections here 🙂

File: src/lib/controllers/aligned-controller.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (balmas) [reply src/lib/controllers/aligned-controller.js:27]: Well, I think the workflow requirements could be different depending upon whether this is upon initial creation of the alignment, or upon adding a translation text after alignment has started.  I think currently the code doesn't support the latter, but I have noted elsewhere in the review that it should and could be done as a separate item. So yes, we could put this requirement on that item :-)

**Comment by balmas** [comment]
> > @balmas , @kirlat - I have worked through all notes, thank you for the attentive look to my code :)
> > Do you think I could merge?
> 
> I have no objections here

Me neither. Please go ahead and merge. Thank you!

**Comment by irina060981** [comment]
great :)
